src/dialogs/create-file.tsx

```diff
@@ -0,0 +1,232 @@
+
+import cockpit from 'cockpit';
+import { FormHelper } from 'cockpit-components-form-helper';
+import type { Dialogs, DialogResult } from 'dialogs';
+import { useInit } from 'hooks';
+import { superuser } from 'superuser';
+
+import "./editor.scss";
+import { useFilesContext } from '../app.tsx';
+import { checkFilename } from '../common.ts';
+import { get_owner_candidates } from '../ownership.tsx';
+
+import { edit_file } from './editor.tsx';
+
+const _ = cockpit.gettext;
+
+async function create_file(filename: string, content?: string, owner?: string) {
+    const file = cockpit.file(filename, { superuser: "try" });
+
+    try {
+        await file.replace("");
+    } catch (err) {
+        console.warn("Cannot create new file", err);
+        throw err;
+    }
+
+    if (owner !== undefined) {
+        try {
+            await cockpit.spawn(["chown", owner, filename], { superuser: "require" });
+        } catch (err) {
+            console.warn("Cannot chown new file", err);
+            throw err;
+        }
+    }
+
+    if (content) {
+        // We need to obtain the tag to retain file ownership
+        // Can't do this async because we can't get the tag via await
+        await cockpit.file(filename, { superuser: "try" }).read()
+                .then(((async (_content: string, tag: string) => {
+                    try {
+                        await cockpit.file(filename, { superuser: "try" }).replace(content, tag);
+                    } catch (err) {
+                        console.warn("Cannot set initial file text", err);
+                        throw err;
+                    }
+                }) as any /* eslint-disable-line @typescript-eslint/no-explicit-any */));
+    }
+}
+
+const CreateFileModal = ({ dialogResult, path } : {
+    dialogResult: DialogResult<string>,
+    path: string
+}) => {
+    const [filename, setFilename] = React.useState<string>("");
+    const [owner, setOwner] = React.useState<string | undefined>();
+    const [initialText, setInitialText] = React.useState<string>("");
+    const [filenameError, setFileNameError] = React.useState<string | null>(null);
+    const [createError, setCreateError] = React.useState<string | null>(null);
+    const [candidates, setCandidates] = React.useState<string[]>([]);
+
+    const { cwdInfo } = useFilesContext();
+
+    useInit(() => {
+        cockpit.user().then(user => {
+            const owner_candidates = [];
+            if (superuser.allowed && cwdInfo) {
+                owner_candidates.push(...get_owner_candidates(user, cwdInfo));
+                setOwner(owner_candidates[0]);
+                setCandidates(owner_candidates);
+            }
+        });
+    });
+
+    const handleEscape = (event: KeyboardEvent) => {
+        if (filename !== "" || initialText !== "") {
+            event.preventDefault();
+        } else {
+            dialogResult.reject(new Error("no file created"));
```

Comment:
Bit weird, this isn't an error. `.result(null)` may make things simpler (see previous thread)